show floating toolbox on button click

Simons102 opened the pull request:
Pop up the Floating Toolbox (if configured) by pressing a button on your pen without tapping the screen.

Even though this seems to work fine, I am not sure if this is the right way to do this, and open to suggestions how it could be done better.

Closes #2901

personalizedrefrigerator commented on src/gui/PageView.cpp:
```diff
@@ -371,6 +371,33 @@ auto XojPageView::onButtonPressEvent(const PositionInputData& pos) -> bool {
+
+    ToolHandler* h = control->getToolHandler();
+
+    if (h->getToolType() == TOOL_FLOATING_TOOLBOX) {
+        gint wx = 0, wy = 0;
+        GtkWidget* widget = xournal->getWidget();
+        gtk_widget_translate_coordinates(widget, gtk_widget_get_toplevel(widget), 0, 0, &wx, &wy);
+
+        wx += std::lround(pos.x) + this->getX();
+        wy += std::lround(pos.y) + this->getY();
+
+        control->getWindow()->floatingToolbox->show(wx, wy);
+    }
```
This code is the same as code in `XojPageView::onButtonPressEvent` (and also `XojPageView::onButtonReleaseEvent`). I think it makes sense to add a private function to `PageView.cpp`:
```c++
/**
 * Shows the floating toolbox at the location of an input event.
 */
static void showFloatingToolbox(XournalView* xournal, const PositionInputData& pos) {
    Control* control = xournal->getControl();

    gint wx = 0, wy = 0;
    GtkWidget* widget = xournal->getWidget();
    gtk_widget_translate_coordinates(widget, gtk_widget_get_toplevel(widget), 0, 0, &wx, &wy);

    wx += std::lround(pos.x) + this->getX();
    wy += std::lround(pos.y) + this->getY();

    control->getWindow()->floatingToolbox->show(wx, wy);
}
```

...that we can use to reduce code duplication!

personalizedrefrigerator commented on src/gui/PageView.h:
```diff
@@ -152,6 +152,11 @@ class XojPageView: public Redrawable, public PageListener {
+    /**
+     * for popup windows like the floating toolbox
+     */
+    bool onButtonClickEvent(const PositionInputData& pos);
```
**Documentation:**
If the comment in `StylusInputHandler` on lines 46-60 is applied, this documentation comment should describe the difference between `onButtonClickEvent` and `onButtonPressEvent`, `onButtonReleaseEvent` (e.g. `this event fires even when no input sequence is actively running`).

---

**Design:**

For future programmers trying to add additional click functionality, I think it could be confusing that `onButtonClick` is only called by `StylusInputHandler` (while other events like `onButtonPressEvent` are also used in, for example, `PenInputHandler`).

Do you think it would make sense for `onButtonClickEvent` to also be called in `onButtonReleaseEvent`?
  * If (in StylusInputHandler) we only call `onButtonClickEvent` directly when there's no input sequence running, we won't be double-calling `onButtonClickEvent`.
  * Additionally, this would let us remove some of the duplicated code I commented about in `PageView`.

personalizedrefrigerator reviewed (commented):
Thank you for working on this! I have a few design comments (mostly about `onButtonClickEvent`). It will be very nice to have this!

personalizedrefrigerator commented:
Thank you for the pull request! This change should make Xournal++ much easier to use.

rolandlo commented:
While this feature is certainly nice to have, I'd like to point out that there are several related issues #2901, #3251, #1939, #1877 (and most likely others) that may have to be addressed at once. It's not only the floating toolbox, but also other tools that should be usable when clicking without touching the screen.

Simons102 commented:
Thanks for the detailed review @personalizedrefrigerator :)
Totally missed that return statement in `onButtonClickEvent`.
By the way, is there any specific reason to use `auto` as the return type?
I added `XojPageView::showFloatingToolbox`, although not as a static function because of the calls to `this->getX()` and `this->getY()`, and removed the check if we're using the floating toolbox from `StylusInputHandler::handleImpl` as you suggested.
I agree with you that `onButtonClickEvent` only being called by `StylusInputHandler` could cause confusion; maybe a good idea would be to also call it wherever pressing buttons on a drawing tablet is handled (which I can't test as I only own a convertible with stylus input), or to give it a different name?
As for calling `onButtonClickEvent` in `onButtonReleaseEvent` too: I'm not sure if I understood your suggestion correctly, but think that a 'click' should either happen when pressing a button or releasing it, and not both on different occasions.
Again, thank you for taking the time take a look at this!
@rolandlo, I'm not sure if some of those Issues are out of the scope of this pull request (for example #1939), as to me the floating toolbox seems more like a one-shot thing, compared to keeping a tool active while a stylus button is pressed. I think the best thing is to either decide on a tool by tool basis which tools should be activated (e.g. hand) while pressing a button and which shouldn't (like for example the pen or highlighter, which would start drawing on the screen), or add an option to the settings where this behaviour can be configured.
I could try to implement one of those and open another pull request though, as I currently have some time to spare :)

personalizedrefrigerator commented:
> As for calling onButtonClickEvent in onButtonReleaseEvent too: I'm not sure if I understood your suggestion correctly, but think that a 'click' should either happen when pressing a button or releasing it, and not both on different occasions.

I'll have to test this, but I think that if <b>onButtonReleaseEvent</b> is called (currently), then <b>onButtonClickEvent</b> will not be called for that event.

If we make <b>onButtonReleaseEvent</b> call <b>onButtonClickEvent</b> for "click like" logic, we have that tools that always start an input sequence when clicking (e.g. mouse) will call <b>onButtonClickEvent</b> in addition to the stylus.
 * If I'm right that this won't lead to calling <b>onButtonClickEvent</b> twice for the same input sequence (e.g. mouse down, mouse move, mouse up is an input sequence), this change could make it easier to add click actions in the future.

Simons102 commented:
While it seems like this doesn't call onButtonClickEvent twice for the same input sequence, it leaves us with this behaviour on mouse down, mouse move, mouse up:

https://user-images.githubusercontent.com/43406797/127057975-51a93a6f-db23-4e87-b6cf-dbb85278e6a6.mp4

which feels a little weird.

It might be a good idea to rename`onButtonClickEvent`, as it's called when a button was pressed and not when a button was pressed and released again, which is more of a complete click. But I have no idea what it could be renamed to.

Would calling `onButtonClickEvent` from `onButtonPressEvent` instead of `onButtonReleaseEvent` and moving the code that handles tools that respond to click actions from `onButtonPressEvent` to `onButtonClickEvent` be another option?
Although I can't think of another tool aside from the floating toolbox that does this.
That would preserve the current behaviour for mouse based input, (most likely) not call `onButtonClickEvent` twice for the same input sequence and not make more than one call to e.g. `FloatingToolbox::show` for the same event.

Simons102 commented:
This should be ready now.

personalizedrefrigerator commented:
I'm noticing that when I use the floating toolbox as I would normally (stylus button down, touch pen to screen, stylus button up, pen above screen), any subsequent taps (with or without the button down) open the floating toolbox.

personalizedrefrigerator commented:
> This should be ready now.

Thank you for working on this! It looks like it's almost ready! See the issue I mentioned above.

Simons102 commented:
Weirdly enough I can only reproduce this on Wayland and not on Xorg.
I don't think that this PR caused that bug, just that it made the bug easier to notice, because it's also possible to get into that state in the current release by pushing the button, touching the stylus to the screen (will open the floating toolbox), removing the stylus from the screen, touching the stylus to the screen (on the floating toolbox), releasing the button and then removing the stylus from the screen again.

Calling `inputHandler->onButtonReleaseEvent` when calling `XojPageView::onButtonClickEvent` from `StylusInputHandler::handleImpl` seems impossible because `inputHandler` was always `null` in that situation.

The issue is that `StylusInputHandler::handleImpl` isn't called for the button release event that is fired when the stylus button is released while the stylus is touching the floating toolbox. That in turn causes `StylusInputHandler::setPressedState` to not be called which means that in `StylusInputHandler` `this->modifier2` or `this->modifier3` simply stay true, and thus `StylusInputHandler` thinks that the stylus button is still being pressed.

I'll try to fix that, but where that `ButtonReleaseEvent` goes is currently beyond me.

personalizedrefrigerator commented:
> I'll try to fix that, 

Fixing it could go in a different pull request. 
This PR looks good to me!

LittleHuba commented:
Please also fix the last mentioned issue. This could lead to buggy input behavior which I do not want to introduce to the development branch.

Simons102 commented:
I don't think that I can fix the last mentioned issue.
That behaviour can be reproduced on the current 1.1.0 release of Xournal++, but so far I could only trigger it on Wayland (Gnome).
It seems like Wayland doesn't give us a button release event for a stylus button outside of the widget that received the corresponding button press event, if the stylus tip was lifted from the screen after pressing and before releasing the button.
I guess that's more likely something that can only properly be fixed upstream.
Nevertheless I created #3302.

Simons102 commented:
If we do something like
```c++
g_signal_connect(this->window, "button-release-event", G_CALLBACK(onButtonReleaseCallback), this);
```
in `MainWindow::MainWindow`, and then in `onButtonReleaseCallback` filter for the device class `INPUT_DEVICE_PEN` or `INPUT_DEVICE_ERASER`, and only in that case pass the event on to `InputContext::handle` (or maybe directly to the `Stylus Handler`), the mentioned issue seems to mostly vanish, but can still be triggered by opening a dropdown from the floating toolbox and then releasing the stylus button.
But doing that causes InputContext::handle to be called twice for a button release event when the stylus is over the document area, and seems kind of hacky.
The issue still remains that we cannot know if a stylus button was released while the stylus was outside of the application's window.

LittleHuba commented:
Okay I thought this was a result of your work. As it already appears in 1.1.0 as well I have no more reason to block this PR.

LittleHuba commented:
Would you mind checking for a ticket over at GTK? If you find none, feel free to create one and reference it with #3302.

Simons102 commented:
Alright, I didn't find a ticket for this at GTK and will open one as soon as I can reproduce this with one of the [demos](https://gitlab.gnome.org/GNOME/gtk/-/tree/master/demos), or have a small and self contained example ready that I can attach to it. Which might take a while.

Simons102 commented:
There now exists a [ticket at GTK](https://gitlab.gnome.org/GNOME/gtk/-/issues/4157), but I'm not sure how I should go about referencing it with #3302, other than putting a link to it in the comments.
Hopefully that actually is an issue with GTK and can be fixed soon.

personalizedrefrigerator commented:
Overall, this PR (for me) makes it easier to use the floating toolbox on Wayland. I would like to merge it in 48 hours, if that seems appropriate.
